src/traversal/writeRelayUpdatePayload.js

```diff
@@ -381,25 +374,29 @@ function handleRangeAdd(
     ));
   }
 
-  if (isOptimisticUpdate) {
-    // optimistic updates need to record the generated client ID for
-    // a to-be-created node
-    RelayMutationTracker.putClientIDForMutation(
-      nodeID,
-      clientMutationID
-    );
-  } else {
-    // non-optimistic updates check for the existence of a generated client
-    // ID (from the above `if` clause) and link the client ID to the actual
-    // server ID.
-    const clientNodeID =
-      RelayMutationTracker.getClientIDForMutation(clientMutationID);
-    if (clientNodeID) {
-      RelayMutationTracker.updateClientServerIDMap(
-        clientNodeID,
-        nodeID
+  const clientMutationID = getString(payload, CLIENT_MUTATION_ID);
+  // subscriptions won't have a clientMutationID so never optimistically add
+  if (clientMutationID) {
```

Comment:
How about changing the outer `if` condition to `if (operation instanceof RelayQuery.Mutation)`, and move the invariant on `clientMutationId` inside the if block? This will get the intended effect of not running the invariant for subscriptions, while keeping the validation for mutations.